### PR TITLE
WIP: Documentation cleanup

### PR DESCRIPTION
**Pull request by rpgoldman**
I've been overwhelmed by the number of Sphinx errors and warnings.  Here are changes to docstrings and configuration to get rid of a bunch of them.

Still a WIP: I'm going to add a bunch of fixes to RST files, too.

### CODE REVIEW

**Comment by rpgoldman** [comment]
@ColCarroll Please review now -- I have done as much as I can do here in the short term.  If you like things you can just merge (feel free to squash merge).  If not, LMK, and I will try to fix things up.

File: pymc3/distributions/timeseries.py [hunk 1]
```diff
@@ -316,7 +316,7 @@ class EulerMaruyama(distribution.Continuous):
     sde_fn : callable
         function returning the drift and diffusion coefficients of SDE
     sde_pars : tuple
-        parameters of the SDE, passed as *args to sde_fn
+        parameters of the SDE, passed as `*args` to `sde_fn`
```
Comment (lucianopaz) [inline pymc3/distributions/timeseries.py:319]:
```suggestion
        parameters of the SDE, passed as ``*args`` to ``sde_fn``
```

File: pymc3/sampling.py [hunk 2]
```diff
@@ -106,7 +106,7 @@ def assign_step_methods(model, step=None, methods=STEP_METHODS,
         A fully-specified model object
     step : step function or vector of step functions
         One or more step functions that have been assigned to some subset of
-        the model's parameters. Defaults to None (no assigned variables).
+        the model's parameters. Defaults to `None` (no assigned variables).
```
Comment (lucianopaz) [inline pymc3/sampling.py:109]:
```suggestion
        the model's parameters. Defaults to ``None`` (no assigned variables).
```

File: pymc3/sampling.py [hunk 3]
```diff
@@ -205,7 +205,7 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None, trace=N
     ----------
     draws : int
         The number of samples to draw. Defaults to 500. The number of tuned samples are discarded
-        by default. See discard_tuned_samples.
+        by default. See `discard_tuned_samples`.
```
Comment (lucianopaz) [inline pymc3/sampling.py:208]:
```suggestion
        by default. See ``discard_tuned_samples``.
```

File: pymc3/sampling.py [hunk 4]
```diff
@@ -218,7 +218,7 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None, trace=N
         * adapt_diag : Start with a identity mass matrix and then adapt a diagonal based on the
           variance of the tuning samples. All chains use the test value (usually the prior mean)
           as starting point.
-        * jitter+adapt_diag : Same as `adapt_diag`, but add uniform jitter in [-1, 1] to the
+        * jitter+adapt_diag : Same as `adapt_diag`\, but add uniform jitter in [-1, 1] to the
```
Comment (lucianopaz) [inline pymc3/sampling.py:221]:
```suggestion
        * jitter+adapt_diag : Same as ``adapt_diag``, but add uniform jitter in [-1, 1] to the
```

File: pymc3/sampling.py [hunk 5]
```diff
@@ -234,26 +234,28 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None, trace=N
         If 'ADVI', number of iterations, if 'nuts', number of draws.
     start : dict, or array of dict
         Starting point in parameter space (or partial point)
-        Defaults to trace.point(-1)) if there is a trace provided and model.test_point if not
+        Defaults to `trace.point(-1))` if there is a trace provided and model.test_point if not
```
Comment (lucianopaz) [inline pymc3/sampling.py:237]:
```suggestion
        Defaults to ``trace.point(-1))`` if there is a trace provided and ``model.test_point`` if not
```

File: pymc3/sampling.py [hunk 6]
```diff
@@ -234,26 +234,28 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None, trace=N
...
         (defaults to empty dict). Initialization methods for NUTS (see `init` keyword) can
-        overwrite the default. For 'SMC' it should be a list of dict with length `chains`.
+        overwrite the default. For 'SMC' step method, `start` should be a list of dicts
```
Comment (lucianopaz) [inline pymc3/sampling.py:239]:
```suggestion
        overwrite the default. For 'SMC' step method, ``start`` should be a list of dicts
```

File: pymc3/sampling.py [hunk 7]
```diff
@@ -234,26 +234,28 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None, trace=N
...
+        of length = `chains`.
```
Comment (lucianopaz) [inline pymc3/sampling.py:240]:
```suggestion
        of length = ``chains``.
```

File: pymc3/sampling.py [hunk 8]
```diff
@@ -234,26 +234,28 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None, trace=N
...
     chains : int
         The number of chains to sample. Running independent chains is important for some
         convergence statistics and can also reveal multiple modes in the posterior. If `None`,
         then set to either `cores` or 2, whichever is larger. For SMC the number of chains is the
         number of draws.
     cores : int
         The number of chains to run in parallel. If `None`, set to the number of CPUs in the
-        system, but at most 4 (for 'SMC' ignored if `pm.SMC(parallel=False)`. Keep in mind that
+        system, but at most 4.  When using 'SMC', this parameter will be ignored if running with
+        `pm.SMC(parallel=False)`. Keep in mind that
```
Comment (lucianopaz) [inline pymc3/sampling.py:258]:
```suggestion
        ``pm.SMC(parallel=False)``. Keep in mind that
```

File: pymc3/sampling.py [hunk 9]
```diff
@@ -271,7 +273,7 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None, trace=N
     discard_tuned_samples : bool
         Whether to discard posterior samples of the tune interval. Ignored when using 'SMC'
     compute_convergence_checks : bool, default=True
-        Whether to compute sampler statistics like gelman-rubin and effective_n.
+        Whether to compute sampler statistics like Gelman-Rubin and `effective_n`.
```
Comment (lucianopaz) [inline pymc3/sampling.py:276]:
```suggestion
        Whether to compute sampler statistics like Gelman-Rubin and ``effective_n``.
```

File: pymc3/sampling.py [hunk 10]
```diff
@@ -1320,22 +1321,18 @@ def sample_prior_predictive(samples=500,
     model : Model (optional if in `with` context)
     vars : Iterable[str]
         A list of names of variables for which to compute the posterior predictive
-         samples.
-        Defaults to `model.named_vars`.
-        DEPRECATED - Use `var_names` instead.
+        samples.  *DEPRECATED* - Use `var_names` argument instead.
```
Comment (lucianopaz) [inline pymc3/sampling.py:1324]:
```suggestion
        samples.  *DEPRECATED* - Use ``var_names`` argument instead.
```

File: pymc3/sampling.py [hunk 11]
```diff
@@ -1320,22 +1321,18 @@ def sample_prior_predictive(samples=500,
...
     var_names : Iterable[str]
         A list of names of variables for which to compute the posterior predictive
-         samples.
-        Defaults to `model.named_vars`.
-
+        samples. Defaults to `model.named_vars`.
```
Comment (lucianopaz) [inline pymc3/sampling.py:1327]:
```suggestion
        samples. Defaults to ``model.named_vars``.
```

File: pymc3/step_methods/smc.py [hunk 12]
```diff
@@ -58,36 +58,38 @@ class SMC:
         It should be between 0 and 1.
     parallel : bool
         Distribute computations across cores if the number of cores is larger than 1
-        (see pm.sample() for details). Defaults to True.
+        (see `pm.sample()` for details). Defaults to True.
```
Comment (lucianopaz) [inline pymc3/step_methods/smc.py:61]:
```suggestion
        (see ``pm.sample()`` for details). Defaults to True.
```

File: pymc3/step_methods/smc.py [hunk 13]
```diff
@@ -58,36 +58,38 @@ class SMC:
...
-     5. Obtain S_{w} by re-sampling according to W.
+     5. Obtain :math:`S_{w}` by re-sampling according to W.
      6. Use W to compute the covariance for the proposal distribution.
      7. For stages other than 0 use the acceptance rate from the previous stage to estimate the
-        scaling of the proposal distribution and n_steps.
-     8. Run N Metropolis chains (each one of length n_steps), starting each one from a different
-        sample in S_{w}.
-     9. Repeat from step 3 until \beta \ge 1.
-    10. The final result is a collection of N samples from the posterior.
+        scaling of the proposal distribution and `n_steps`.
```
Comment (lucianopaz) [inline pymc3/step_methods/smc.py:88]:
```suggestion
        scaling of the proposal distribution and ``n_steps``.
```

File: pymc3/step_methods/smc.py [hunk 14]
```diff
@@ -58,36 +58,38 @@ class SMC:
...
+     8. Run N Metropolis chains (each one of length `n_steps`), starting each one from a different
```
Comment (lucianopaz) [inline pymc3/step_methods/smc.py:89]:
```suggestion
     8. Run N Metropolis chains (each one of length ``n_steps``), starting each one from a different
```

File: pymc3/backends/base.py [hunk 15]
```diff
@@ -256,6 +256,17 @@ class MultiTrace:
     For any methods that require a single trace (e.g., taking the length
     of the MultiTrace instance, which returns the number of draws), the
     trace with the highest chain number is always used.
+
+    Attributes
+    ----------
+        nchains : int
+            Number of chains in the `MultiTrace`.
```
Comment (lucianopaz) [inline pymc3/backends/base.py:263]:
```suggestion
            Number of chains in the ``MultiTrace``.
```

File: pymc3/backends/base.py [hunk 16]
```diff
@@ -256,6 +256,17 @@ class MultiTrace:
...
+        chains : `List[int]`
```
Comment (lucianopaz) [inline pymc3/backends/base.py:264]: I'm not sure if type hints should be rendered as code or just emphasized

```suggestion
        chains : ``List[int]``
```

File: pymc3/backends/base.py [hunk 17]
```diff
@@ -256,6 +256,17 @@ class MultiTrace:
...
+            List of chain indices
+        report : str
+            Report on the sampling process.
+        varnames : `List[str]`
```
Comment (lucianopaz) [inline pymc3/backends/base.py:268]: Same as the above, not sure if type hints should be rendered as code or simply emphasized

```suggestion
        varnames : ``List[str]``
```

File: pymc3/backends/base.py [hunk 18]
```diff
@@ -363,19 +374,23 @@ def stat_names(self):
...
+        """Add variables to traces.
 
         Parameters
         ----------
         vals : dict (str: array-like)
              The keys should be the names of the new variables. The values are expected to be
-             array-like object. For traces with more than one chain the length of each value
-             should match the number of total samples already in the trace (chains * iterations),
+             array-like objects. For traces with more than one chain the length of each value
+             should match the number of total samples already in the trace `(chains * iterations)`,
```
Comment (lucianopaz) [inline pymc3/backends/base.py:385]:
```suggestion
             should match the number of total samples already in the trace ``(chains * iterations)``,
```

File: pymc3/backends/base.py [hunk 19]
```diff
@@ -363,19 +374,23 @@ def stat_names(self):
...
              otherwise a warning is raised.
         overwrite : bool
             If `False` (default) a ValueError is raised if the variable already exists.
             Change to `True` to overwrite the values of variables
+
+        Returns
+        -------
+            Nothing.
```
Comment (lucianopaz) [inline pymc3/backends/base.py:393]: Technically, it returns `None`.

```suggestion
            ``None``
```

File: pymc3/data.py [hunk 20]
```diff
@@ -141,8 +141,8 @@ class Minibatch(tt.TensorVariable):
     if we want 1d slice of size 10 we do
     >>> x = Minibatch(data, batch_size=10)
 
-    Note, that your data is cast to `floatX` if it is not integer type
-    But you still can add `dtype` kwarg for :class:`Minibatch`
+    Note that your data is cast to `floatX` if it is not integer type
```
Comment (lucianopaz) [inline pymc3/data.py:144]:
```suggestion
    Note that your data is cast to ``floatX`` if it is not integer type
```

File: pymc3/data.py [hunk 21]
```diff
@@ -141,8 +141,8 @@ class Minibatch(tt.TensorVariable):
...
+    But you still can add the `dtype` kwarg for :class:`Minibatch`
```
Comment (lucianopaz) [inline pymc3/data.py:145]:
```suggestion
    But you still can add the ``dtype`` kwarg for :class:`Minibatch`
```

File: pymc3/distributions/continuous.py [hunk 22]
```diff
@@ -4060,7 +4060,7 @@ def logcdf(self, value):
         References
         ----------
         .. [Machler2012] Martin Mächler (2012).
-            "Accurately computing log(1-exp(-|a|)) Assessed by the Rmpfr
+            "Accurately computing :math:  `\log(1-\exp(- \mid a \mid<))` Assessed by the Rmpfr
```
Comment (lucianopaz) [inline pymc3/distributions/continuous.py:4063]:
```suggestion
            "Accurately computing :math:`\log(1-\exp(- \mid a \mid<))` Assessed by the Rmpfr
```

**Comment by ColCarroll** [comment]
Super helpful as always @rpgoldman!

**Review by lucianopaz — APPROVED** [review]
Incredible job @rpgoldman!! I just left some minor suggestions on switching markdown inline code to the one used in rst's, but maybe you can ignore them if the `recommonmark` understands that it is dealing with markdown instead of an reStructuredText.

**Comment by rpgoldman** [comment]
@lucianopaz Thanks for the corrections.  I will make them and push the relevant ones after testing with the documentation generator.

**Comment by rpgoldman** [comment]
@lucianopaz You are right -- I didn't understand the markdown interaction.  Fixing and will push.

**Pull request merged**